library/core/src/fmt/mod.rs

```diff
@@ -244,34 +494,28 @@ impl<W: Write + ?Sized> Write for &mut W {
 #[stable(feature = "rust1", since = "1.0.0")]
 #[rustc_diagnostic_item = "Formatter"]
 pub struct Formatter<'a> {
-    flags: u32,
-    fill: char,
-    align: rt::Alignment,
-    width: Option<usize>,
-    precision: Option<usize>,
+    options: FormattingOptions,
```

Comment:
> This makes a Formatter object bigger than before

Are you sure? Before, `flags` was a `u32`, so 4 bytes. With each flag in its own field, there are two bools and two small enums wrapped into `Option` – both should be one byte per field, so 4 bytes in total as well. Unless I'm not seeing something here, both the original code and the new implementation have the same size.

However, the benchmark shows a performance regression, and when I realized that `Formatter` didn't get bigger and therefore, `flags` may not be the culprit, I was already halfway through refactoring `FormattingOptions` to use a bitmask for `flags` (like it did before). So, either way, let's test that, it might help even if a difference in size is not the cause of the performance regression. Could you please rerun the benchmark?

Reply:
> Are you sure?

No :)

> Before, flags was a u32, so 4 bytes. With each flag in its own field, there are two bools and two small enums wrapped into Option – both should be one byte per field, so 4 bytes in total as well.

Good point!

I think that in the near future we might want to put more things in the fields flag though. At least the alignment can easiliy fit in there, and the Some/None flag of the width and precision could fit in there as well:

```rust
pub struct FormattingOptions {
    flags: u32, // sign, zero pad, alt, debug-as-hex, width flag, precision flag, alignment
    fill: char,
    width: usize, // only used if width flag is set.
    precision: usize, // only used if width flag is set.
}
```

But maybe that's overkill.

Reply:
> ```rust
> pub struct FormattingOptions {
>     flags: u32, // sign, zero pad, alt, debug-as-hex, width flag, precision flag, alignment
>     fill: char,
>     width: usize, // only used if width flag is set.
>     precision: usize, // only used if width flag is set.
> }
> ```

well, `fill` has lots of spare bits, you could probably merge it with `flags`, where `flags` is just the upper bits and `fill` is the lower 21 bits. though this would only be beneficial on `<= 32`-bit systems unless `width` and `precision` were changed to be `u32` or the struct was `#[repr(packed(4))]`.

Reply:
I agree that there is definitely potential for further optimization here. I'll leave the solution using `flags` – while it is a bit harder to read, the possible performance benefits in the future seem more important to me.

The benchmark still reports a small performance regression (~0.1 %). While I don't think this is a big problem, it would still be nice if this PR would not introduce any performance regressions. To make the code more maintainable, I changed all internal accesses to formatting options to use the getters/setters on `FormattingOptions` instead of direct field access. I suspect this is at least a major factor in the performance regression. I have reverted this now, can you please rerun the benchmark @m-ou-se?